tensorrt_llm/_torch/modules/fused_moe/fused_moe_cutlass.py

```diff
@@ -0,0 +1,640 @@
+            elif self.reduce_results:
+                outputs = self.all_reduce(inputs)
+        return outputs
+
+    def forward_chunk(
+        self,
+        x: Union[torch.Tensor, Fp4QuantizedTensor],
+        router_logits: torch.Tensor,
+        cutlass_min_latency_mode: bool = False,
+        output_dtype: Optional[torch.dtype] = None,
+        all_rank_num_tokens: Optional[List[int]] = None,
+        use_dp_padding: Optional[bool] = None,
+    ) -> torch.Tensor:
+        if isinstance(x, Fp4QuantizedTensor):
+            assert output_dtype is not None
+            output_dtype = output_dtype
+        else:
+            output_dtype = x.dtype
+
+        use_fp8_block_scaling = False
+        use_w4a8_group_scaling = False
+        weight_dtype = self.w3_w1_weight.dtype
+
+        token_selected_experts, token_final_scales = self.routing_method.apply(
+            router_logits)
+        if self.balancer_layer is None:
+            token_selected_slots = token_selected_experts
+        else:
+            # If attention DP is enabled, token_selected_experts is a local rank tensor,
+            # so we need to offset the round robin position by ep_rank
+            token_selected_slots = self.balancer_layer.route(
+                token_selected_experts, offset_by_ep_rank=self.use_dp)
+
+        # If load balancer is disabled, the statistics are collected from expert IDs.
+        # If load balancer is enabled, the statistics are collected from expert slot IDs.
+        ExpertStatistic.set_layer(self.layer_idx)
+        ExpertStatistic.maybe_add_info(self.num_slots, token_selected_slots)
+
+        assert token_selected_slots.shape[
+            1] == self.routing_method.experts_per_token
+        assert token_selected_slots.shape == token_final_scales.shape
+        assert token_selected_slots.shape[0] == router_logits.shape[0]
+        assert token_final_scales.dtype == torch.float32
+        assert token_selected_slots.dtype == torch.int32
+
+        if self.apply_router_weight_on_input:
+            assert self.routing_method.top_k == 1, "Current workaround only supports top-1 routing"
+            assert x.dtype != torch.float8_e4m3fn, "Current workaround for apply_router_weight_on_input does not support fp8 input"
+            x = x * token_final_scales.to(x.dtype)
+            # TODO: remove this once we have correct fusedmoe kernel ready
+            token_final_scales = None
+
+        token_count = x.shape[0]
+
+        alltoall_info = None
+
+        if self.enable_alltoall:
+            x, token_selected_slots, token_final_scales, alltoall_info = \
+                self.alltoall_prepare_maybe_dispatch(all_rank_num_tokens,
+                                                     x,
+                                                     token_selected_slots,
+                                                     token_final_scales)
+
+        x_sf = None
+        if self.has_any_quant:
```

Comment:
Maybe it doesn't belong in this refactor, but should we have a generic method/class `ActivationQuantizer` function that does this quantization and is controlled by the base quant method?
I don't exactly know how to make this correctly generic, and may require changes to the moe op

Reply:
I thought about it and gave up the idea in the first refactor. We sometimes calls quantize and sometimes don't. It's hard to model this behavior correctly. I'll revisit this later.